packages/structures/src/bitfields/BitField.ts

```diff
@@ -0,0 +1,205 @@
+import type { EnumLike, RecursiveReadonlyArray } from '../utils/types';
+
+/**
+ * Data that can be resolved to give a bitfield. This can be:
+ * A bit number (this can be a number literal or a value taken from {@link BitField.Flags})
+ * A string bit number
+ * An instance of BitField
+ * An Array of BitFieldResolvable
+ */
+export type BitFieldResolvable<Flags extends string, Type extends bigint | number> =
+	| Flags
+	| Readonly<BitField<Flags, Type>>
+	| RecursiveReadonlyArray<Flags | Readonly<BitField<Flags, Type>> | Type | `${bigint}`>
+	| Type
+	| `${bigint}`;
+
+/**
+ * Data structure that makes it easy to interact with a bitfield.
+ */
+export abstract class BitField<Flags extends string, Type extends bigint | number = number> {
+	/**
+	 * Numeric bitfield flags.
+	 * <info>Defined in extension classes</info>
+	 */
+	public static readonly Flags: EnumLike<unknown, bigint | number> = {};
+
+	protected static readonly DefaultBit: bigint | number = 0;
+
+	/**
+	 * Bitfield of the packed bits
+	 */
+	public bitfield: Type;
+
+	/**
+	 * @param bits - Bit(s) to read from
+	 */
+	public constructor(
+		bits: BitFieldResolvable<Flags, Type> = (this.constructor as typeof BitField<Flags, Type>).DefaultBit as Type,
```

Comment:
pro tip: you can define `['constructor']: typeof BitField<Flags, Type>` on the class level to skip all this casting, or make a getter function for it (https://www.typescriptlang.org/play/?#code/MYGwhgzhAECiAeYC2AHEBTAPAdQPYHdp14AXdAOwBMYByMG6AH2hoCMaA+aAbwCgBISulBgATumgQSYEgEtgRRKgwAFUbhQAuaCVEBXdAG5eAoSPHQA2jWC5yU-cBK5RNALraSATxTpcAMzglNCw8fA5jU2FwC29fSgA1MBADbTDI-n89cgAKAEoeAX5ZQJySAAtZCAA6OPRE5INoAF5WlnoCvn5uiqrq23tdPScXauJkELUNIv4AelnoRaXllYA9AH4igF8TbvnoAEEQMlFyGVkAN3Rd-l6akdEx4NV1FCL9laWNgR2BAHN0CRoA98oVuuISHpTjpKvc7A5hs5RNBIDofH5AggJhgcAQIj9eFsgA)

Reply:
Initially I came up with
```ts
type NonAbstract<Constructor> = Constructor extends abstract new (...args: infer Args) => infer Instance ? new (...args: Args) => Instance : never; 
```

but that got rid of static props, but with some tweaking I got this:
https://www.typescriptlang.org/play/?#code/C4TwDgpgBAcg9gOwIICMDOwBOBDAxsAHgBUoIAPYCBAEzSm3Sz2CgQgHcoAKAOj+0wBzNAC56CEAEooAXgB84kApkAoKFBLlKNOgww58rDtz48BwsQEsEAMwiYoSIWmnyo1uw4CSCDNgS4EGrqUAD83GycvPzOYk7Crgo+fgEQ0gBkUAAKlrgA1sQANFB5ECBwNhpywepibABu9gDcKq16TIa4ADbYaHQAomTYALZgXRAEAOpwnFpUtFAA5NiLUAA+SyiLCgDeKgCQ1BDdAtB+wLmkQ6PjWZhwYGJYAK4QLQdHJ5jQANqLuIh9M98HBMIsALpieDIRgGQigSAVKCDEZjCbTdhyOTvQ7HHrfKAIiDUABq2C6rzEGJxNmeCC40j2+32lkqXGAAAtLGgeETSeTXrIZDIlitGQdmZzuTwAb4XiDMDxyKjbvcwBKAL6tZmRQlcnmyoEKhktfZajVAA